File: scripts/sign_artifact.py
#!/usr/bin/env python3
"""
sign_artifact.py — Universal Artifact Signature generator
Produces: UUIDv7 + ISO-8601 timestamp + SHA-256 + YAML frontmatter patch

Author:    Quang Cuong Huynh (Registered IP Authority)
Authority: Registered — Quang Cuong Huynh
Usage:
    # Sign a single file (patch frontmatter in-place)
    python scripts/sign_artifact.py --path docs/myfile.md

    # Sign and print provenance JSON (no file modification)
    python scripts/sign_artifact.py --path docs/myfile.md --dry-run

    # Batch sign all unsigned files in a directory
    python scripts/sign_artifact.py --batch --dir personas/roles/

    # Generate standalone signature record
    python scripts/sign_artifact.py --path docs/myfile.md --provenance
"""
import argparse, hashlib, json, os, re, sys, time
from datetime import datetime, timezone
from pathlib import Path


# ─── UUIDv7 (RFC 9562) ────────────────────────────────────────────────────────
def uuidv7() -> str:
    ms = int(time.time() * 1000)
    rand = int.from_bytes(os.urandom(10), "big")
    uuid_int = (
        (ms << 80)
        | (0x7 << 76)
        | ((rand >> 62) << 64)
        | (0b10 << 62)
        | (rand & 0x3FFFFFFFFFFFFFFF)
    )
    h = f"{uuid_int:032x}"
    return f"{h[:8]}-{h[8:12]}-{h[12:16]}-{h[16:20]}-{h[20:]}"


# ─── SHA-256 ──────────────────────────────────────────────────────────────────
def sha256_file(path: Path) -> str:
    """Hash file content excluding the YAML frontmatter block."""
    raw = path.read_text(encoding="utf-8")
    # Strip frontmatter (--- ... ---) for stable hash across registration runs
    body = re.sub(r"^---\n.*?\n---\n", "", raw, flags=re.DOTALL)
    return hashlib.sha256(body.encode("utf-8")).hexdigest()


def sha256_str(s: str) -> str:
    return hashlib.sha256(s.encode("utf-8")).hexdigest()


# ─── ISO-8601 timestamp ───────────────────────────────────────────────────────
def iso_now() -> str:
    return datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ")


# ─── Frontmatter patch ────────────────────────────────────────────────────────
def patch_frontmatter(path: Path, artifact_id: str, sha256: str) -> bool:
    """Patch artifact_id and sha256 fields in YAML frontmatter. Returns True if changed."""
    text = path.read_text(encoding="utf-8")
    if not text.startswith("---"):
        print(f"  ⚠ No frontmatter: {path}", file=sys.stderr)
        return False

    changed = False
    # Patch artifact_id if empty
    if re.search(r'^artifact_id:\s*""', text, re.MULTILINE):
        text = re.sub(r'^artifact_id:\s*""', f'artifact_id: "{artifact_id}"', text, flags=re.MULTILINE)
        changed = True
    # Patch sha256 if empty
    if re.search(r'^sha256:\s*""', text, re.MULTILINE):
        text = re.sub(r'^sha256:\s*""', f'sha256: "{sha256}"', text, flags=re.MULTILINE)
        changed = True

    if changed:
        path.write_text(text, encoding="utf-8")
    return changed


# ─── Provenance record ────────────────────────────────────────────────────────
def make_provenance(path: Path, artifact_id: str, sha256: str, agent: str = "register_artifact.py") -> dict:
    return {
        "artifact_id":      artifact_id,
        "file_path":        str(path),
        "created_by_agent": agent,
        "supervised_by":    "Quang Cuong Huynh",
        "outputs_hash":     f"sha256:{sha256}",
        "timestamp":        iso_now(),
        "signatures": [{"type": "timestamp", "signer": "sign_artifact.py", "fingerprint": sha256[:16]}]
    }


# ─── CLI ──────────────────────────────────────────────────────────────────────
def main():
    ap = argparse.ArgumentParser(description="Universal Artifact Signature generator")
    ap.add_argument("--path",       help="Single file to sign")
    ap.add_argument("--batch",      action="store_true", help="Sign all .md/.yml/.yaml/.json in --dir")
    ap.add_argument("--dir",        default=".", help="Directory for batch mode (default: .)")
    ap.add_argument("--dry-run",    action="store_true", help="Print actions, no file writes")
    ap.add_argument("--provenance", action="store_true", help="Write .provenance.json alongside file")
    ap.add_argument("--agent",      default="sign_artifact.py", help="Agent ID for provenance record")
    args = ap.parse_args()

    targets = []
    if args.path:
        targets = [Path(args.path)]
    elif args.batch:
        d = Path(args.dir)
        targets = list(d.rglob("*.md")) + list(d.rglob("*.yml")) + list(d.rglob("*.yaml")) + list(d.rglob("*.json"))
        # Skip schema files themselves, node_modules, .git
        targets = [t for t in targets if ".git" not in str(t) and "node_modules" not in str(t) and "schema.json" not in t.name]
    else:
        ap.print_help()
        sys.exit(1)

    total, signed, skipped = 0, 0, 0
    for p in targets:
        if not p.is_file():
            continue
        total += 1
        text = p.read_text(encoding="utf-8", errors="replace")
        # Only process files with frontmatter containing empty artifact_id
        if not (text.startswith("---") and 'artifact_id: ""' in text):
            skipped += 1
            continue

        uid    = uuidv7()
        digest = sha256_file(p)
        prov   = make_provenance(p, uid, digest, args.agent)

        if args.dry_run:
            print(f"[DRY-RUN] {p}")
            print(f"  artifact_id : {uid}")
            print(f"  sha256      : {digest}")
        else:
            changed = patch_frontmatter(p, uid, digest)
            if changed:
                signed += 1
                print(f"  ✓ SIGNED  {p}  →  {uid[:18]}...")
                if args.provenance:
                    prov_path = p.with_suffix(".provenance.json")
                    prov_path.write_text(json.dumps(prov, indent=2), encoding="utf-8")
                    print(f"  ✓ PROV    {prov_path}")
            else:
                skipped += 1

    if not args.dry_run:
        print(f"\nSummary: {signed} signed | {skipped} skipped | {total} total")


if __name__ == "__main__":
    main()

File: scripts/sign_artifact.js
#!/usr/bin/env node
/**
 * sign_artifact.js — Universal Artifact Signature generator (JavaScript/Node.js)
 * Author:    Quang Cuong Huynh (Registered IP Authority)
 * Authority: Registered — Quang Cuong Huynh
 *
 * Mirrors sign_artifact.py for Node.js / Claude Code / frontend toolchain environments.
 *
 * Usage:
 *   node scripts/sign_artifact.js --path docs/myfile.md
 *   node scripts/sign_artifact.js --batch --dir personas/
 *   node scripts/sign_artifact.js --path docs/myfile.md --dry-run
 *   node scripts/sign_artifact.js --generate-uuid          # just print a UUIDv7
 *   node scripts/sign_artifact.js --generate-hash --path f  # just SHA-256
 */
const fs   = require('fs');
const path = require('path');
const crypto = require('crypto');

// ─── UUIDv7 (RFC 9562) ──────────────────────────────────────────────────────
function uuidv7() {
  const ms     = BigInt(Date.now());
  const randBuf = crypto.randomBytes(10);
  const rand   = BigInt('0x' + randBuf.toString('hex'));
  const uuid_int =
    (ms << 80n)
    | (0x7n << 76n)
    | ((rand >> 62n) << 64n)
    | (0b10n << 62n)
    | (rand & 0x3FFFFFFFFFFFFFFFn);
  const h = uuid_int.toString(16).padStart(32, '0');
  return `${h.slice(0,8)}-${h.slice(8,12)}-${h.slice(12,16)}-${h.slice(16,20)}-${h.slice(20)}`;
}

// ─── SHA-256 ──────────────────────────────────────────────────────────────────
function sha256Body(filePath) {
  const raw  = fs.readFileSync(filePath, 'utf8');
  const body = raw.replace(/^---\n[\s\S]*?\n---\n/, '');
  return crypto.createHash('sha256').update(body, 'utf8').digest('hex');
}

function sha256String(s) {
  return crypto.createHash('sha256').update(s, 'utf8').digest('hex');
}

// ─── ISO-8601 timestamp ───────────────────────────────────────────────────────
function isoNow() {
  return new Date().toISOString().replace(/\.\d{3}Z$/, 'Z');
}

// ─── Frontmatter patch ────────────────────────────────────────────────────────
function patchFrontmatter(filePath, artifactId, digest) {
  let text = fs.readFileSync(filePath, 'utf8');
  let changed = false;
  if (text.includes('artifact_id: ""')) {
    text = text.replace('artifact_id: ""', `artifact_id: "${artifactId}"`);
    changed = true;
  }
  if (text.includes('sha256: ""')) {
    text = text.replace('sha256: ""', `sha256: "${digest}"`);
    changed = true;
  }
  if (changed) fs.writeFileSync(filePath, text, 'utf8');
  return changed;
}

// ─── Recursive file scan ──────────────────────────────────────────────────────
function scanDir(dir, exts = ['.md', '.yml', '.yaml']) {
  const results = [];
  function walk(d) {
    for (const entry of fs.readdirSync(d, { withFileTypes: true })) {
      const full = path.join(d, entry.name);
      if (entry.isDirectory() && entry.name !== '.git' && entry.name !== 'node_modules') {
        walk(full);
      } else if (entry.isFile() && exts.includes(path.extname(entry.name))) {
        results.push(full);
      }
    }
  }
  walk(dir);
  return results;
}

// ─── CLI ──────────────────────────────────────────────────────────────────────
const args = process.argv.slice(2);
const getArg = (flag) => { const i = args.indexOf(flag); return i !== -1 ? args[i+1] : null; };
const hasFlag = (flag) => args.includes(flag);

if (hasFlag('--generate-uuid')) {
  const n = parseInt(getArg('--n') || '1');
  for (let i = 0; i < n; i++) console.log(uuidv7());
  process.exit(0);
}

if (hasFlag('--generate-hash')) {
  const p = getArg('--path');
  if (!p) { console.error('--path required'); process.exit(1); }
  console.log(sha256Body(p));
  process.exit(0);
}

if (hasFlag('--timestamp')) {
  console.log(isoNow());
  process.exit(0);
}

const filePath  = getArg('--path');
const batchMode = hasFlag('--batch');
const dirPath   = getArg('--dir') || '.';
const dryRun    = hasFlag('--dry-run');

let targets = [];
if (filePath)   targets = [filePath];
else if (batchMode) targets = scanDir(dirPath);
else { console.log('Usage: node sign_artifact.js --path <file> | --batch [--dir <dir>] [--dry-run]'); process.exit(0); }

let signed = 0, skipped = 0;
for (const t of targets) {
  if (!fs.existsSync(t)) continue;
  const text = fs.readFileSync(t, 'utf8');
  if (!text.startsWith('---') || !text.includes('artifact_id: ""')) { skipped++; continue; }

  const uid    = uuidv7();
  const digest = sha256Body(t);

  if (dryRun) {
    console.log(`[DRY] ${t}`);
    console.log(`  artifact_id : ${uid}`);
    console.log(`  sha256      : ${digest}`);
  } else {
    const changed = patchFrontmatter(t, uid, digest);
    if (changed) {
      signed++;
      console.log(`  ✓ SIGNED  ${t}  →  ${uid.slice(0,18)}...`);
    } else {
      skipped++;
    }
  }
}

if (!dryRun) console.log(`\nSigned: ${signed} | Skipped: ${skipped} | Total: ${targets.length}`);
